src/file-upload/dropzone.tsx

```diff
@@ -0,0 +1,113 @@
+import InternalBox from '../box/internal';
+
+interface DropzoneProps {
+  onChange: (files: File[]) => void;
+  children: React.ReactNode;
+}
+
+export function useDropzoneVisible() {
+  const [isDropzoneVisible, setDropzoneVisible] = useState(false);
+
+  // Registering global drag events listeners.
+  useEffect(() => {
+    // The timer helps avoiding dropzone blinking.
+    let dragTimer: null | ReturnType<typeof setTimeout> = null;
+
+    // The file-upload dropzone is shown when the user drags files over the to the browser.
+    const onDocumentDragOver = (event: DragEvent) => {
+      event.preventDefault();
+
+      if (event.dataTransfer && event.dataTransfer.types && event.dataTransfer.types.indexOf('Files') !== -1) {
+        setDropzoneVisible(true);
+        dragTimer && clearTimeout(dragTimer);
+      }
+    };
+
+    // When the files are no longer dragged over the browser the state must be reset.
+    const onDocumentDragLeave = (event: DragEvent) => {
+      event.preventDefault();
+
+      dragTimer = setTimeout(() => setDropzoneVisible(false), 25);
+    };
+
+    // If the files were dropped the state must be reset.
+    const onDocumentDrop = (event: DragEvent) => {
+      event.preventDefault();
+
+      dragTimer = setTimeout(() => setDropzoneVisible(false), 25);
+    };
+
+    document.addEventListener('dragover', onDocumentDragOver, false);
+    document.addEventListener('dragleave', onDocumentDragLeave, false);
+    document.addEventListener('drop', onDocumentDrop, false);
+
+    return () => {
+      dragTimer && clearTimeout(dragTimer);
+      document.removeEventListener('dragover', onDocumentDragOver);
+      document.removeEventListener('dragleave', onDocumentDragLeave);
+      document.removeEventListener('drop', onDocumentDrop);
+    };
+  }, []);
+
+  return isDropzoneVisible;
+}
+
+export function Dropzone({ onChange, children }: DropzoneProps) {
+  const dropzoneRef = useRef<HTMLDivElement>(null);
+  const [isDropzoneHovered, setDropzoneHovered] = useState(false);
+
+  const onDragOver = (event: React.DragEvent) => {
+    event.preventDefault();
+    event.stopPropagation();
+    setDropzoneHovered(true);
+
+    if (event.dataTransfer) {
+      event.dataTransfer.dropEffect = 'copy';
+    }
+  };
+
+  const onDragLeave = (event: React.DragEvent) => {
+    event.preventDefault();
+    event.stopPropagation();
+    setDropzoneHovered(false);
+
+    if (event.dataTransfer) {
+      event.dataTransfer.dropEffect = 'none';
+    }
+  };
+
+  const onDrop = (event: React.DragEvent) => {
+    event.preventDefault();
+
+    const newFiles: File[] = [];
+    const nullableFiles = event.dataTransfer.items
+      ? Array.from(event.dataTransfer.items).map(item => (item.kind === 'file' && item.getAsFile()) || null)
+      : Array.from(event.dataTransfer.files);
+    nullableFiles.forEach(fileOrNull => fileOrNull && newFiles.push(fileOrNull));
+
+    onChange(newFiles);
+    setDropzoneHovered(false);
```

Comment:
minor: I would call `setDropzoneHovered` before firing the event. Otherwise, if an error happens in the event handler, the state would not get updated